src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs

```diff
@@ -413,49 +426,148 @@ public static SequenceValidity PeekFirstSequence(ReadOnlySpan<byte> data, out in
             return SequenceValidity.Incomplete;
         }
 
-        public static void EscapeString(ReadOnlySpan<char> value, Span<char> destination, int indexOfFirstByteToEscape, out int written)
+        private static void EscapeString(ReadOnlySpan<char> value, Span<char> destination, JavaScriptEncoder encoder, ref int written)
         {
-            Debug.Assert(indexOfFirstByteToEscape >= 0 && indexOfFirstByteToEscape < value.Length);
+            // todo: issue #39523: add an Encode(ReadOnlySpan<char>) decode API to System.Text.Encodings.Web.TextEncoding to avoid utf16->utf8->utf16 conversion.
 
-            value.Slice(0, indexOfFirstByteToEscape).CopyTo(destination);
-            written = indexOfFirstByteToEscape;
-            int consumed = indexOfFirstByteToEscape;
+            Debug.Assert(encoder != null);
 
-            while (consumed < value.Length)
+            // Convert char to byte.
+            byte[] utf8DestinationArray = null;
+            Span<byte> utf8Destination;
+            int length = checked((value.Length) * JsonConstants.MaxExpansionFactorWhileTranscoding);
+            if (length > JsonConstants.StackallocThreshold)
             {
-                char val = value[consumed];
-                if (NeedsEscaping(val))
+                utf8DestinationArray = ArrayPool<byte>.Shared.Rent(length);
+                utf8Destination = utf8DestinationArray;
+            }
+            else
+            {
+                unsafe
                 {
-                    EscapeNextChars(value, val, destination, ref consumed, ref written);
+                    byte* ptr = stackalloc byte[JsonConstants.StackallocThreshold];
+                    utf8Destination = new Span<byte>(ptr, JsonConstants.StackallocThreshold);
```

Comment:
Since this isn't a ref struct, or being passed to a ref struct, you should be able to avoid the `unsafe` by declaring the original variable `Span<byte> utf8Destination = stackalloc byte[0];`, then here you can just declare it as `utf8Destination = stackalloc byte[JsonConstants.StackallocThreshold];`.

Or, just declare it originally as a stackalloc of size threshold, and only optionally overwrite it.

Reply:
Thanks. If #39523 is implemented for 3.0 then this code will just go away. If not implemented, then yes I will make the change to avoid `unsafe` here.